packages/modules/src/hooks/useScrollDepth.ts

```diff
@@ -0,0 +1,37 @@
+import { useLayoutEffect, useRef } from 'react';
+
+export function useScrollDepth(
+    effect: (scrollPercent: number) => void,
+    deps: React.DependencyList,
+    debounce?: number,
+): void {
+    const pageHeight = useRef(document.documentElement.scrollHeight - window.innerHeight);
+
+    let throttleTimeout: NodeJS.Timeout | null = null;
+
+    function onScroll() {
+        const currentPosition = window.scrollY;
+        const percentScroll = (currentPosition / pageHeight.current) * 100;
+        effect(percentScroll);
+        throttleTimeout = null;
+    }
+
+    useLayoutEffect(() => {
+        function handleScroll() {
+            if (debounce) {
+                if (throttleTimeout === null) {
+                    throttleTimeout = setTimeout(onScroll, debounce);
```

Comment:
Do these `setTimeouts` ever need to be cleared? If the user scrolls a bunch before the first timeout and then stops, there’ll be several scheduled calls to `onScroll` which will all do the same thing, right? I’m not sure it matters, but I wanted to ask.

Reply:
The point of this condition is to only create _one_ scheduled call of `onScroll` during the debounce period, and not to schedule it again until after it's been executed, as `onScroll` resets `throttleTimeout` to `null`.

It's not about clearing the timeouts- you can't do that without calling `clearTimeout` anyway- it's about not creating more than we need in the first place.

Reply:
Ah yes, I misunderstood the logic with setting `throttleTimeout` and only setting a new timeout if it’s null – thanks for the explanation.